src/frontend/src/utils/condition.rs

```diff
@@ -290,24 +290,29 @@ impl Condition {
 
             // analyze exprs in the group. scan_range is not updated
             for expr in group.clone() {
-                if let Some((input_ref, lit)) = expr.as_eq_literal() {
+                if let Some((input_ref, const_expr)) = expr.as_eq_const() &&
+                    let Ok(const_expr) = const_expr.cast_implicit(input_ref.data_type) {
                     assert_eq!(input_ref.index, order_column_ids[i]);
-                    if lit.is_null() {
+                    let Ok(Some(value)) = const_expr.eval_row_const() else {
+                        // column = NULL or failed to eval
                         return always_false();
-                    }
-                    let lit = lit.eval_as(input_ref.data_type);
-                    if !eq_conds.is_empty() && eq_conds.into_iter().all(|l| l != lit) {
+                    };
+                    if !eq_conds.is_empty() && eq_conds.into_iter().all(|l| l != value) {
                         return always_false();
                     }
-                    eq_conds = vec![lit];
-                } else if let Some((input_ref, in_lit_list)) = expr.as_in_literal_list() {
+                    eq_conds = vec![value];
+                } else if let Some((input_ref, in_const_list)) = expr.as_in_const_list() {
                     assert_eq!(input_ref.index, order_column_ids[i]);
                     let mut scalars = vec![];
-                    for lit in in_lit_list.into_iter().unique() {
-                        if lit.is_null() {
+                    for const_expr in in_const_list.into_iter().unique() {
```

Comment:
`unique` here is less effective, as it compares the expression rather than the evaluated result.